docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,125 @@
+
+## Dependencies
+
+- the name of the Kubernetes Secret needs to be known (configured in the
+  StorageClass, or somewhere else)
+- each tenant (Kubernetes namespace) should use the same name for the Secret
+  that contains the Vault token
+- the Ceph-CSI sidecar needs to store the namespace (and name of the Secret?)
+  together with the metadata of the PVC
+
+## Implementation Outline
+
+- when creating the PVC the Ceph-CSI provisioner needs to store the namespace
+  (and Kubernetes Secret?) of the PVC in its metadata
+  - stores the `csi.volume.owner` (name of Tenant) in the metadata of the
+    volume and sets it as `rbdVolume.Owner`
+- the Ceph-CSI node-plugin needs to request the token in the NodeStage CSI
+  operation and create/get the key for the PVC
+- the Ceph-CSI provisioner needs to request the token to delete the key for the
+  PVC in the VolumeDelete CSI operation
+
+### Fetching Tenant Tokens from Kubernetes
+
+CSI is an abstraction and should not communicate with Kubernetes or other
+Container Orchestration systems directly. Therefore, it is needed to
+communicate with a service that can provide the token from the tenants. This
+service can be provided by a sidecar, exposing an endpoint as a
+UNIX-Domain-Socket to communicate through.
+
+This sidecar does not exist at the moment. There are other features within
+Ceph-CSI that will benefit from this sidecar, e.g. Topology support. Because
+Ceph-CSI already uses a Kubernetes client to fetch details from Kubernetes, it
+should be acceptable to fetch the Vault Token configuration for the tenants the
+same way.
+
+## Configuration Details
+
+- the current KMS configuration file needs extensions for Vault Token support
+- introduce a new KMS-type: VaultTokenKMS (the current VaultKMS uses a
+  Kubernetes ServiceAccount)
+- configuration of the VaultTokenKMS can be very similar to VaultKMS for common
+  settings
+- the configuration can override the defaults for each tenant separately
+  - Vault Service connection details (address, TLS options, ...)
+  - default token (in case of a single token for every tenant)
+  - name of the Kubernetes Secret that can be looked up per tenant
+- tokens for each tenant can be added to the configuration (insecure & needs
+  admin involvement)
+- the configuration points to a Kubernetes Secret per tenant that contains the
+  token, but also can contain alternative options for the VaultTokenKMS service
+
+### Example of the KMS configuration file for Vault Tokens
+
+The configuration is available in the Ceph-CSI containers as
+`/etc/ceph-csi-encryption-kms-config/config.json`:
+
+```json
+{
+    "vault-with-tokens": {
+        "encryptionKMSType": "vaulttoken",
+        "vaultAddress": "http://vault.default.svc.cluster.local:8200",
+        "vaultCAFromSecret": "vault-ca",
+        "vaultCAVerify": "false",
+        "vaultToken": "@@B64_ENCODED_DEFAULT_TOKEN@@",
+        "tenantConfigName": "@@SECRET_NAME_IN_NAMESPACE@@",
+        "tenants": {
```

Comment:
Do we need this flexibility at all? I.e each tenant token as part of a common store in the ceph-csi namespace.

I ask, because if each tenant has a different token, gathering them in the same store seems counter productive to the security requirement for which independent tokens are created in the first place.

Technically as the ceph-csi namespace is given read access to the secrets it can be argued that the user/admin of the ceph-csi namespace can anyway read all the secrets.

Reply:
Tokens will most likely not be stored in the configuration file. That would be ideal. Having the storage admin (that maintains the Ceph-CSI namespace) update/add tokens from tenants is extremely painful. Tenants may have different Vault services that they need to use, so having a per-tenant option to overload options will be helpful.

Reply:
Per-tenant option to overload vault service is fine. Thanks for taking care of not adding tokens in the config map :+1: